### ISSUE TITLE
Integer types mixup

### ISSUE THREAD

**Issue by mikk-leini-krakul** [libjxl/libjxl#1534]
**Describe the bug**
We are trying to test JXL on ARM Cortex-M7 embedded microcontroller using Arm GNU Toolchain for bare-metal target.  No idea if it would work and how well, but first step is to make it compiling. Unfortunately the libjxl code does not compile because integer types are mixed up. E.g. int and int32_t cross-used.

An example code segment from "enc_chroma_from_luma.cc":
```c
//  Function header
void ComputeDC(const ImageF& dc_values, bool fast, int* dc_x, int* dc_b)

// Calling function
void CfLHeuristics::ComputeDC(bool fast, ColorCorrelationMap* cmap) {
  int32_t ytob_dc = 0;
  int32_t ytox_dc = 0;
  HWY_DYNAMIC_DISPATCH(ComputeDC)(dc_values, fast, &ytox_dc, &ytob_dc);
```
Error:
`libjxl/lib/jxl/enc_chroma_from_luma.cc:345:62: error: invalid conversion from 'int32_t*' {aka 'long int*'} to 'int*' [-fpermissive]`

Another example with STL vectors in enc_ans.cc:
```c
// Function header:
std::vector<int> CreateFlatHistogram(int length, int total_count);

// Calling code (3 relevant lines)
using ANSHistBin = int32_t;
// ...
std::vector<ANSHistBin> counts(histogram, histogram + alphabet_size);
// ...
counts = CreateFlatHistogram(alphabet_size, ANS_TAB_SIZE);
```
Error:
`libjxl/lib/jxl/enc_ans.cc:492:61: error: no match for 'operator=' (operand types are 'std::vector<long int>' and 'std::vector<int>')`

Prrinting out relevant GCC intrinsic macros with:
`echo | arm-none-eabi-gcc -E -dM - `
Reveals that:
`#define __INT32_TYPE__ long int`

The sizeof(int) and sizeof(int32_t) matches, both are 4 (bytes), but semantics are different and code does not compile.

We started making local fixes, but new issues keep popping up. Did not find a way to solve it with compiler flags either.

**To Reproduce**
There is no official support for ARM Cortex-M target building so it is fully manual process to create/modify the build system for it. I skip the steps, it doesn't matter.

**Expected behavior**
I am not 100% sure that libjxl violates C/C++ standard(s), but it would make very much sense to avoid mixing types and use width-defined types only (int32_t, uint32_t, int64_t, etc.). It might also be possible to use some static code analyzer to detect the issues.

**Environment**
 - OS: Windows
 - Compiler version:[ Arm GNU Toolchain](https://developer.arm.com/Tools%20and%20Software/GNU%20Toolchain) 11.2-2022.02 AArch32 bare-metal target (arm-none-eabi).
 - CPU type: ARM Cortex M7 (32-bits)
 - cjxl/djxl version string: N/A

**Additional context**

**Comment by mo271** [comment]
I completely agree, thank you very much for the detailed bug report, @mikk-leini-krakul!
I think we should switch to width-defined types throughout.

**Comment by mikk-leini-krakul** [comment]
Thanks @mo271 ! We tried regex replace across whole libjxl from "(unsigned) int" to "(u)int32_t" and the other way, but it caused new problems in both cases. So we actually went on with the manual approach and solved all compilation errors by changing from int to int32_t. I could create a pull request, but without knowing the algorithm better it feels too risky. There could be some integer under or overflows. The code still doesn't work either - there are other challenges to solve. Compilation errors took couple of hours to fix, so not very big deal after all and maybe it's better if JXL team does it.

I should probably create separate issue, but there's just one more small incompatibility issue with arm-none-eabi:
Linker has trouble with std::atomic<uint64_t> types in cache_aligned.cc. It doesn't link since it can't find atomic functions for 8-byte types (e.g. __atomic_fetch_add_8). Since those types were used for statistics only, we just left them out and didn't find a proper solution.

**Comment by mo271** [comment]
> Thanks @mo271 ! We tried regex replace across whole libjxl from "(unsigned) int" to "(u)int32_t" and the other way, but it caused new problems in both cases. So we actually went on with the manual approach and solved all compilation errors by changing from int to int32_t. I could create a pull request, but without knowing the algorithm better it feels too risky. There could be some integer under or overflows. The code still doesn't work either - there are other challenges to solve. Compilation errors took couple of hours to fix, so not very big deal after all and maybe it's better if JXL team does it.
> 

A pull request would always be welcome, even if it still has some test failures! We would probably be doing the same thing, changing int to int32_t.

> I should probably create separate issue, but there's just one more small incompatibility issue with arm-none-eabi: Linker has trouble with std::atomic<uint64_t> types in cache_aligned.cc. It doesn't link since it can't find atomic functions for 8-byte types (e.g. __atomic_fetch_add_8). Since those types were used for statistics only, we just left them out and didn't find a proper solution.

For this a separate issue would be great indeed!

**Comment by mo271** [comment]
Are you planning to make a pull request on this, @mikk-leini-krakul?

**Comment by mikk-leini-krakul** [comment]
@mo271  Sorry for delay. I started a few weeks ago but then went on vacation. I got the static library compiling with minimal changes and created pull request #1640 . As I explained in that pull request - it only solves int types issue because to make it actually run on target, at least I had to cut out few more things (see above). So I can't promise it would work on any bare metal ARM.

For general information:
We got the encoder working on NXP Cortex M7, but results are not satisfactory so we probably won't use it and there will be no further pull requests from me. JXL consumes a lot (in embedded world terms) of dynamic RAM so we couldn't test with large (FHD+) images and it's slow. VGA (320x240) resolution image encoding on single core CM7 at 1 GHz with quality 50 and Brotly effort 7 takes 82 seconds. The code ran from external QSPI using ICACHE because it didn't fit to ITCM or internal RAM, so that made it slower. Regardless of that it seems there's still a lot of overhead in the emulated 128-bit vector math in Highway library. On vector processor all that code would be inlined to few assembly instructions, but on simple CPU, there's a lot of simple arithmetic 5-7 sub-functions downwards (no inlining to spare ROM) . To run JXL on non-vector CPU, a much lighter implementation is needed or some HW acceleration. What's good is that JXL is designed to work on single thread as well, no problems with that.
